src/annotator/config/test/config-test.js

```diff
@@ -0,0 +1,185 @@
+'use strict';
+
+var proxyquire = require('proxyquire');
+
+var fakeSettings = {};
+var fakeSharedSettings = {};
+var configFrom = proxyquire('../config', {
+  './settings': fakeSettings,
+  '../../shared/settings': fakeSharedSettings,
+});
+
+function fakeWindow() {
+  return {
+    document: 'THE_DOCUMENT',
+    location: {href: 'LOCATION_HREF'},
+  };
+}
+
+describe('annotator.config', function() {
+  beforeEach('reset fakeSettings', function() {
+    fakeSettings.iFrameSrc = sinon.stub().returns('IFRAME_URL');
+    fakeSettings.annotations = sinon.stub().returns(null);
+    fakeSettings.query = sinon.stub().returns(null);
+    fakeSettings.configFuncSettingsFrom = sinon.stub().returns({});
+  });
+
+  beforeEach('reset fakeSharedSettings', function() {
+    fakeSharedSettings.jsonConfigsFrom = sinon.stub().returns({});
+  });
+
+  context("when there's an application/annotator+html <link>", function() {
+    it("returns the <link>'s href as config.app", function() {
+      assert.equal(configFrom(window).app, 'IFRAME_URL');
+    });
+  });
+
+  context("when there's no application/annotator+html <link>", function() {
+    beforeEach('remove the application/annotator+html <link>', function() {
+      fakeSettings.iFrameSrc.returns(null);
+    });
+
+    it('sets config.app to null', function() {
+      var config = configFrom(fakeWindow());
+
+      assert.isNull(config.app);
+    });
+  });
+
+  it('gets the JSON settings from the document', function() {
+    var window_ = fakeWindow();
+
+    configFrom(window_);
+
+    assert.calledOnce(fakeSharedSettings.jsonConfigsFrom);
+    assert.calledWithExactly(
+      fakeSharedSettings.jsonConfigsFrom, window_.document);
+  });
+
+  context('when jsonConfigsFrom() returns a non-empty object', function() {
+    it('reads the setting into the returned config', function() {
+      // configFrom() just blindly adds any key: value settings that
+      // jsonConfigsFrom() returns into the returns options object.
+      fakeSharedSettings.jsonConfigsFrom.returns({foo: 'bar'});
+
+      var config = configFrom(fakeWindow());
+
+      assert.equal(config.foo, 'bar');
+    });
+  });
+
+  it('gets config settings from window.hypothesisConfig()', function() {
+    var window_ = fakeWindow();
+
+    configFrom(window_);
+
+    assert.calledOnce(fakeSettings.configFuncSettingsFrom);
+    assert.calledWithExactly(fakeSettings.configFuncSettingsFrom, window_);
+  });
+
+  context('when configFuncSettingsFrom() returns an object', function() {
+    it('reads arbitrary settings from configFuncSettingsFrom() into config', function() {
+      fakeSettings.configFuncSettingsFrom.returns({foo: 'bar'});
+
+      var config = configFrom(fakeWindow());
+
+      assert.equal(config.foo, 'bar');
+    });
+
+    specify('hypothesisConfig() settings override js-hypothesis-config ones', function() {
```

Comment:
All this overriding behaviour is going to be disappearing from `config.js` into `settings.js` functions (much of it is untested currently)